### ISSUE TITLE
Invalid type in JSON write (__NSDate)

### ISSUE THREAD

**Issue by sergeh** [katzer/cordova-plugin-local-notifications#496]
When updating a notification I get this error

    *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'Invalid type in     JSON write (__NSDate)'
    *** First throw call stack:
    (0x2774f5f7 0x34e89c77 0x2774f53d 0x284c4e69 0x284c69ef 0x2774ef6d 0x27660078 0x2774c991 0x284c6671 0x284c4da7 0x284c6919 0x27676071 0x27675f99 0x284c6861 0x284c4de5 0x284c6919 0x2767af9b 0x276740d7 0x284c6861 0x284c4de5 0x284c4bbd 0x284c5a5f 0xedebf 0xda823 0xeef7f 0xbcdb7 0x2519cb 0x25b7bb 0x25c881 0x35567e25 0x35567b78)
    libc++abi.dylib: terminating with uncaught exception of type NSException

and here's how I update the scheduled notification

    window.plugin.notification.local.update({
            id:         alarm.id, 
            text:       message,
            data:       { novenas: alarm.data.novenas, day: alarm.nextDay}
          });

`alarm.data.novenas` is a string and `alarm.nextDay` is an integer 

Setting the notification works fine, but updating it always results in that error... any ideas what's going on?

**Comment by katzer** [comment]
Please check if the object does not contain null values. String, Integer, undefined,... is ok, but not null.

**Comment by nexus-uw** [comment]
@sergeh, have you resolved this issue? My app keeps crashing due to the same issue. 

@katzer The source of the exception appears to originate when trying to serialize the notifications updated on property

I have attached the stack trace from the  crash report below. The important lines are 26 - 28 (IMClient)
```

Crashed Thread:        3  Dispatch queue: com.apple.root.default-qos

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000

Application Specific Information:
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'Invalid type in JSON write (__NSDate)'
terminating with uncaught exception of type NSException
abort() called
CoreSimulator 117.15 - Device: iPhone 6 - Runtime: iOS 8.3 (12F69) - DeviceType: iPhone 6

Application Specific Backtrace 1:
0   CoreFoundation                      0x00249746 __exceptionPreprocess + 182
1   libobjc.A.dylib                     0x027e2a97 objc_exception_throw + 44
2   CoreFoundation                      0x0024966d +[NSException raise:format:] + 141
3   Foundation                          0x024e84d5 _writeJSONValue + 808
4   Foundation                          0x024ece82 ___writeJSONObject_block_invoke + 225
5   CoreFoundation                      0x002487d9 ____NSDictionaryEnumerate_block_invoke421 + 41
6   CoreFoundation                      0x00133d42 CFBasicHashApply + 130
7   CoreFoundation                      0x0016f91e __NSDictionaryEnumerate + 622
8   CoreFoundation                      0x0016f62d -[NSDictionary enumerateKeysAndObjectsWithOptions:usingBlock:] + 45
9   CoreFoundation                      0x0016f575 -[NSDictionary enumerateKeysAndObjectsUsingBlock:] + 53
10  Foundation                          0x024eca31 _writeJSONObject + 357
11  Foundation                          0x024e83d3 _writeJSONValue + 550
12  Foundation                          0x024ecd46 ___writeJSONArray_block_invoke + 129
13  CoreFoundation                      0x0016dcc3 __53-[__NSArrayM enumerateObjectsWithOptions:usingBlock:]_block_invoke + 99
14  CoreFoundation                      0x0016d38f -[__NSArrayM enumerateObjectsWithOptions:usingBlock:] + 287
15  CoreFoundation                      0x0016d1e5 -[NSArray enumerateObjectsUsingBlock:] + 53
16  Foundation                          0x024ecc52 _writeJSONArray + 272
17  Foundation                          0x024e842a _writeJSONValue + 637
18  Foundation                          0x024ecd46 ___writeJSONArray_block_invoke + 129
19  CoreFoundation                      0x0018296f __53-[__NSArrayI enumerateObjectsWithOptions:usingBlock:]_block_invoke + 63
20  CoreFoundation                      0x0018289c -[__NSArrayI enumerateObjectsWithOptions:usingBlock:] + 268
21  CoreFoundation                      0x0016d1e5 -[NSArray enumerateObjectsUsingBlock:] + 53
22  Foundation                          0x024ecc52 _writeJSONArray + 272
23  Foundation                          0x024e842a _writeJSONValue + 637
24  Foundation                          0x024e816d -[_NSJSONWriter dataWithRootObject:options:error:] + 125
25  Foundation                          0x024eb0d4 +[NSJSONSerialization dataWithJSONObject:options:error:] + 371
26  IMClient                            0x0005265e -[NSArray(CDVJSONSerializingPrivate) cdv_JSONString] + 94
27  IMClient                            0x0003d0af -[CDVPluginResult argumentsAsJSON] + 319
28  IMClient                            0x0005385e -[CDVCommandDelegateImpl sendPluginResult:callbackId:] + 398
29  IMClient                            0x00021c37 __42-[APPLocalNotification getOptions:byType:]_block_invoke + 807
30  libdispatch.dylib                   0x02e6dd8a _dispatch_call_block_and_release + 15
31  libdispatch.dylib                   0x02e8aecf _dispatch_client_callout + 14
32  libdispatch.dylib                   0x02e75348 _dispatch_root_queue_drain + 745
33  libdispatch.dylib                   0x02e763e7 _dispatch_worker_thread3 + 108
34  libsystem_pthread.dylib             0x031ca1da _pthread_wqthread + 724
35  libsystem_pthread.dylib             0x031c7e2e start_wqthread + 30
 ```

**Comment by prevostc** [comment]
I have the same issues while getting a notification, the error is thrown after this call
https://github.com/katzer/cordova-plugin-local-notifications/blob/master/www/local-notification-util.js#L274

    exec(fn, null, 'LocalNotification', action, params);
    // here action is equal to "getAll" and params is equal to ["2"]

Here is the stack trace

    Apr 27 12:11:41 macbook.local tqf[67799]: *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'Invalid type in JSON write (__NSDate)'
        *** First throw call stack:
        (
            0   CoreFoundation                      0x0066e746 __exceptionPreprocess + 182
            1   libobjc.A.dylib                     0x03b23a97 objc_exception_throw + 44
            2   CoreFoundation                      0x0066e66d +[NSException raise:format:] + 141
            3   Foundation                          0x038294d5 _writeJSONValue + 808
            4   Foundation                          0x0382de82 ___writeJSONObject_block_invoke + 225
            5   CoreFoundation                      0x0066d7d9 ____NSDictionaryEnumerate_block_invoke421 + 41
            6   CoreFoundation                      0x00558d42 CFBasicHashApply + 130
            7   CoreFoundation                      0x0059491e __NSDictionaryEnumerate + 622
            8   CoreFoundation                      0x0059462d -[NSDictionary enumerateKeysAndObjectsWithOptions:usingBlock:] + 45
            9   CoreFoundation                      0x00594575 -[NSDictionary enumerateKeysAndObjectsUsingBlock:] + 53
            10  Foundation                          0x0382da31 _writeJSONObject + 357
            11  Foundation                          0x038293d3 _writeJSONValue + 550
            12  Foundation                          0x0382dd46 ___writeJSONArray_block_invoke + 129
            13  CoreFoundation                      0x00592cc3 __53-[__NSArrayM enumerateObjectsWithOptions:usingBlock:]_block_invoke + 99
            14  CoreFoundation                      0x0059238f -[__NSArrayM enumerateObjectsWithOptions:usingBlock:] + 287
            15  CoreFoundation                      0x005921e5 -[NSArray enumerateObjectsUsingBlock:] + 53
            16  Foundation                          0x0382dc52 _writeJSONArray + 272
            17  Foundation                          0x0382942a _writeJSONValue + 637
            18  Foundation                          0x0382dd46 ___writeJSONArray_block_invoke + 129
            19  CoreFoundation                      0x005a796f __53-[__NSArrayI enumerateObjectsWithOptions:usingBlock:]_block_invoke + 63
            20  CoreFoundation                      0x005a789c -[__NSArrayI enumerateObjectsWithOptions:usingBlock:] + 268
            21  CoreFoundation                      0x005921e5 -[NSArray enumerateObjectsUsingBlock:] + 53
            22  Foundation                          0x0382dc52 _writeJSONArray + 272
            23  Foundation                          0x0382942a _writeJSONValue + 637
            24  Foundation                          0x0382916d -[_NSJSONWriter dataWithRootObject:options:error:] + 125
            25  Foundation                          0x0382c0d4 +[NSJSONSerialization dataWithJSONObject:options:error:] + 371
            26  tqf                            0x0024429e -[NSArray(CDVJSONSerializingPrivate) cdv_JSONString] + 94
            27  tqf                            0x0022ecef -[CDVPluginResult argumentsAsJSON] + 319
            28  tqf                            0x0024549e -[CDVCommandDelegateImpl sendPluginResult:callbackId:] + 398
            29  tqf                            0x0014d5d7 __42-[APPLocalNotification getOptions:byType:]_block_invoke + 807
            30  libdispatch.dylib                   0x04260d8a _dispatch_call_block_and_release + 15
            31  libdispatch.dylib                   0x0427decf _dispatch_client_callout + 14
            32  libdispatch.dylib                   0x04268348 _dispatch_root_queue_drain + 745
            33  libdispatch.dylib                   0x042693e7 _dispatch_worker_thread3 + 108
            34  libsystem_pthread.dylib             0x045bd1da _pthread_wqthread + 724
            35  libsystem_pthread.dylib             0x045bae2e start_wqthread + 30
        )
    Apr 27 12:11:42 macbook com.apple.CoreSimulator.SimDevice.CE6684A9-1854-4492-BABB-7C12D7BB7A5D.launchd_sim[60216] (UIKitApplication:com.it.tqf[0xf4][67799]): Service exited due to signal: Abort trap: 6
    Apr 27 12:11:42 macbook.local SpringBoard[60230]: Application 'UIKitApplication:com.it.tqf[0xf4]' crashed.
    Apr 27 12:11:42 macbook assertiond[60234]: assertion failed: 14D136 12F69: assertiond + 12334 [8E465C14-E4AE-36E9-AFBE-D28459602962]: 0x1

**Comment by katzer** [comment]
Fixed: 0157a9f4507857777640a8a41fbc1eedca744b3e
